packages/zoe/src/contractSupport/atomicTransfer.js

```diff
@@ -0,0 +1,132 @@
+const { Fail, quote: q } = assert;
+
+/**
+ * @typedef {[
+ *   fromSeat?: ZCFSeat,
+ *   toSeat?: ZCFSeat,
+ *   fromAmounts?: AmountKeywordRecord,
+ *   toAmounts?: AmountKeywordRecord
+ * ]} TransferArgs
+ */
+
+/**
+ * TODO Refactor from being a helper into being zcf's replacement for
+ * reallocate. Is currently a helper during the transition, to avoid
+ * interference with progress on Zoe durability.
+ *
+ * See the helpers below, `fromOnly`, `toOnly`, and `atomicTransfer`,
+ * which will remain helpers, for further conveniences for expressing
+ * atomic rearragements clearly.
+ *
+ * @param {ZCF} zcf
+ * @param {TransferArgs[]} transfers
+ */
+export const atomicRearrange = (zcf, transfers) => {
+  const uniqueSeatSet = new Set();
+  for (const [
+    fromSeat = undefined,
+    toSeat = undefined,
+    _subtrahend,
+    _addend,
```

Comment:
awkward names. why destructure them at all? the type makes clear what's in the tuple

Reply:
I don't understand this comment. There are four distinct things but only two types. 

In any case, I find generally find it clearer to destructure the distinct things I need into separate variables.  This might just be a bad habit taught by the weird things we need to do to be defensive in JavaScript, where a property access can have side effects and deliver different values different times. Regarding just the instability, a nice defensive practice is "If you don't want inconsistent answers, ask each question only once."

We designed our Passable types so they would not create such hazards. We're not there yet (need ref to reentrancy vulnerability, cc @mhofman ), but we plan to be, and are programming under the assumption we succeed. So, under this assumption, this defensive practice is not needed for Passables. As of this PR, this is not yet a Passable-only boundary, but it will be by the time we need it to be defensive. So my practice is overkill. But I just like it better anyway.

Fixed the names but still destructuring. Resolving.